crates/typst/src/layout/grid/lines.rs

```diff
@@ -0,0 +1,749 @@
+                };
+                offset += size;
+                interrupted_segment
+            } else {
+                // Reached the end of all tracks, so we interrupt and finish
+                // the current segment.
+                current_segment.take()
+            }
+        },
+    )
+}
+
+/// Returns the correct stroke with which to draw a vline right before column
+/// 'x' when going through row 'y', given the stroke of the user-specified line
+/// at this position, if any.
+/// If the vline would go through a colspan, returns None (shouldn't be drawn).
+/// If the one (when at the border) or two (otherwise) cells to the left and
+/// right of the vline have right and left stroke overrides, respectively,
+/// then the cells' stroke overrides are folded together with the vline's
+/// stroke (with priority to the vline's stroke, followed by the right cell's
+/// stroke, and, finally, the left cell's) and returned. If, however, the cells
+/// around the vline at this row do not have any stroke overrides, then the
+/// vline's own stroke is directly returned.
+pub(super) fn vline_stroke_at_row(
+    grid: &CellGrid,
+    x: usize,
+    y: usize,
+    stroke: Option<Arc<Stroke<Abs>>>,
+) -> Option<Arc<Stroke<Abs>>> {
+    if x != 0 && x != grid.cols.len() {
+        // When the vline isn't at the border, we need to check if a colspan would
+        // be present between columns 'x' and 'x-1' at row 'y', and thus overlap
+        // with the line.
+        // To do so, we analyze the cell right after this vline. If it is merged
+        // with a cell before this line (parent_x < x) which is at this row or
+        // above it (parent_y <= y), this means it would overlap with the vline,
+        // so the vline must not be drawn at this row.
+        let first_adjacent_cell = if grid.has_gutter {
+            // Skip the gutters, if x or y represent gutter tracks.
+            // We would then analyze the cell one column after (if at a gutter
+            // column), and/or one row below (if at a gutter row), in order to
+            // check if it would be merged with a cell before the vline.
+            (x + x % 2, y + y % 2)
+        } else {
+            (x, y)
+        };
+        let Axes { x: parent_x, y: parent_y } = grid
+            .parent_cell_position(first_adjacent_cell.0, first_adjacent_cell.1)
+            .unwrap();
+
+        if parent_x < x && parent_y <= y {
+            // There is a colspan cell going through this vline's position,
+            // so don't draw it here.
+            return None;
+        }
+    }
+
+    let left_cell_stroke = x
+        .checked_sub(1)
+        .and_then(|left_x| grid.parent_cell(left_x, y))
+        .and_then(|left_cell| left_cell.stroke.right.as_ref());
+    let right_cell_stroke = if x < grid.cols.len() {
+        grid.parent_cell(x, y)
+            .and_then(|right_cell| right_cell.stroke.left.as_ref())
+    } else {
+        None
+    };
```

Comment:
Would it make sense to replace the assertions in `CellGrid::entry` with
```rust
if x >= self.cols.len() || y >= self.rows.len() {
    return None;
}
```
Then, we could skip this if here and rely on `parent_cell` to return `None`. Might also simplify other places. On the downside, it might make bad code fail silently instead of panicking.

Reply:
As far as I'm aware, this is the only place where we simultaneously accept `grid.cols.len()` as a valid `x` value and try to check for a cell at `x` (similarly for `y` and `grid.rows.len()` at the other function), so I don't think that'd be worth it overall. We could reconsider this if we depend on this at more places in the future though, as I see it.

Reply:
Okay, I just thought that it might be more idiomatic, but as I said it might also lead silent failure, so maybe it's best to leave it.